core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1116,23 +1120,33 @@ class DAGScheduler(
    */
   private[scheduler] def handleTaskCompletion(event: CompletionEvent) {
     val task = event.task
+    val taskId = task.partitionId
     val stageId = task.stageId
     val taskType = Utils.getFormattedClassName(task)
 
     outputCommitCoordinator.taskCompleted(
       stageId,
-      task.partitionId,
+      taskId,
       event.taskInfo.attemptNumber, // this is a task attempt number
       event.reason)
 
-    // The success case is dealt with separately below, since we need to compute accumulator
-    // updates before posting.
-    if (event.reason != Success) {
-      val attemptId = task.stageAttemptId
-      listenerBus.post(SparkListenerTaskEnd(stageId, attemptId, taskType, event.reason,
-        event.taskInfo, event.taskMetrics))
+    // Reconstruct task metrics. Note: this may be null if the task has failed.
+    val taskMetrics: TaskMetrics =
+    if (event.accumUpdates.nonEmpty) {
+      try {
+        TaskMetrics.fromAccumulatorUpdates(event.accumUpdates)
+      } catch {
+        case NonFatal(e) =>
+          logError(s"Error when attempting to reconstruct metrics for task $taskId", e)
+          null
+      }
+    } else {
+      null
     }
 
+    listenerBus.post(SparkListenerTaskEnd(
```

Comment:
(update: as I commented elsewhere I reverted this part of the change because it was failing tests in unexpected ways)